src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                        });
+                    }
+
+                    if (document.Id == semanticDocument.Document.Id)
+                    {
+                        UpdateExpressionInOriginalFunction(semanticDocument, expressionCopy, oldMethodDeclaration, parameterName, editor, allOccurrences, cancellationToken);
+
+                        var parameterType = semanticDocument.SemanticModel.GetTypeInfo(expressionCopy, cancellationToken).Type ?? semanticDocument.SemanticModel.Compilation.ObjectType;
+                        var refKind = syntaxFacts.GetRefKindOfArgument(expressionCopy);
+                        var parameter = generator.ParameterDeclaration(name: parameterName, type: generator.TypeExpression(parameterType), refKind: refKind);
+                        editor.InsertParameter(oldMethodDeclaration, insertionIndex, parameter);
+                    }
+
+                    var newRoot = editor.GetChangedRoot();
+                    modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(firstCallSite.Id, newRoot);
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// Ties the identifiers within the expression back to their associated parameter
+        /// </summary>
+        public static Dictionary<TIdentifierNameSyntax, IParameterSymbol> MapExpressionToParameters(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var nameToParameterDict = new Dictionary<TIdentifierNameSyntax, IParameterSymbol>();
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = document.SemanticModel;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol != null && symbol is IParameterSymbol parameterSymbol)
+                {
+                    if (!nameToParameterDict.ContainsKey(variable))
+                    {
+                        nameToParameterDict.Add(variable, parameterSymbol);
+                    }
+                }
+            }
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is contained within something that is "parameterized"
+        /// </summary>
+        private (bool isParameterized, bool hasOptionalParameter) ExpressionWithinParameterizedMethod(SemanticDocument semanticDocument,
+                                                                                                      TExpressionSyntax expression,
+                                                                                                      CancellationToken cancellationToken)
+        {
+            var methodSymbol = GetMethodSymbolFromExpression(semanticDocument, expression, cancellationToken);
```

Comment:
this method name is unclear as to what it does.

Reply:
CHanged it to the GetRequiredEnclosingSymbol